src/Microsoft.DotNet.XHarness.Tests.Runners/xUnit/XUnitFilter.cs

```diff
@@ -71,29 +80,176 @@ public static XUnitFilter CreateClassFilter(string className, bool exclude, stri
+            if (string.IsNullOrEmpty(testClassNamespace))
+            {
+                // if we exclude, since we have no namespace, we include the test
+                return log (!Exclude);
+            }
+
+            if (string.Equals(testClassNamespace, SelectorValue, StringComparison.InvariantCultureIgnoreCase))
+            {
+                return log (Exclude);
+            }
+
+            // same logic as with no namespace
+            return log (!Exclude);
+        }
+
+        public bool IsExcluded(TestAssemblyInfo assembly, Action<string>? reportFilteredAssembly = null)
+        {
+            if (FilterType != XUnitFilterType.Assembly)
+                throw new InvalidOperationException("Filter is not targeting assemblies.");
+
+            Func<bool, bool> log = (result) => ReportFilteredAssembly(assembly, result, reportFilteredAssembly);
+
+            if (string.Equals(AssemblyName, assembly.FullPath, StringComparison.Ordinal))
+            {
+                return log(Exclude);
+            }
+
+            string fileName = Path.GetFileName(assembly.FullPath);
+            if (string.Equals(fileName, AssemblyName, StringComparison.Ordinal))
+            {
+                return log (Exclude);
+            }
+
+            // nor path of name matched the filter, therefore return the oposite of the Exclude value
+            return log (!Exclude);
+        }
+
+        public bool IsExcluded(ITestCase testCase, Action<string>? log = null)
+        {
+            Func<bool, bool>? reportFilteredTest = null;
+            if (log != null)
+                reportFilteredTest = (result) => ReportFilteredTest(testCase, result, log);
+            return FilterType switch
+            {
+                XUnitFilterType.Trait => ApplyTraitFilter(testCase, reportFilteredTest),
+                XUnitFilterType.TypeName => ApplyTypeNameFilter(testCase, reportFilteredTest),
+                XUnitFilterType.Single => ApplySingleFilter(testCase, reportFilteredTest),
+                XUnitFilterType.Namespace => ApplyNamespaceFilter(testCase, reportFilteredTest),
+                _ => throw new InvalidOperationException($"Unsupported filter type {FilterType}")
+            };
+        }
+
+        bool ReportFilteredTest(ITestCase testCase, bool excluded, Action<string>? log = null)
+        {
+            const string includedText = "Included";
+            const string excludedText = "Excluded";
+
+            if (log == null)
+                return excluded;
+
+            var selector = FilterType == XUnitFilterType.Trait ?
+                $"'{SelectorName}':'{SelectorValue}'" : $"'{SelectorValue}'";
+
+            log($"[FILTER] {(excluded? excludedText : includedText)} test (filtered by {FilterType}; {selector}): {testCase.DisplayName}");
+            return excluded;
+        }
+
+        bool ReportFilteredAssembly(TestAssemblyInfo assemblyInfo, bool excluded, Action<string>? log = null)
```

Comment:
I don't understand why this returns `excluded` back? For example in the usage above:
```csharp
return log(Excluded);
```

Is is quite unclear what is going on

Reply:
Is a pattern that I inherited.. is not nice, but I did not want to make more changes, is like doing:

```chsarp
if (logFilteredTests)
{
    log.WriteLine ("My log msg");
}
return excluded;
```

mainly, so that we do not need to write that if all the time.